### DISCUSSION TITLE
[Question][Hw4] Complexity of grammar

### DISCUSSION THREAD

**Discussion by park-sunho** [prosyslab-classroom/cs424-program-reasoning#336] [Q&A]
Hello. I have a question about complexity of grammar (for each testcase) in Homework 4. 

Can we assume that, for each testcase, grammar is given as a subset of following rules?
- Start := (+ Start Start), (- Start Start), (ite Bexp Start Start)
- Bexp := (= Start Start), (and Start Start), (< Start Start), (<= Start Start), (> Start Start), (>= Start Start)
- `Start` also contains a finite subset of integers

or, should we consider complex cases where there are more nonterminals other than `Start` and `Bexp`? For instance,
- Start := (+ Start Iexp), (ite Bexp Iexp (- 0 Start)), x, 0
- Bexp := (= Start Start), (and Start Start), (< Start Start), (<= Start Start), (> Start Start), (>= Start Start)
- Iexp := 2, (- Iexp Start)

If it is allowed, then like in this example, we cannot apply many of equivalence reduction rules for top-down approach introduced in class (for example, `if (true) e1 e2` cannot be ignored by considering `e1` only), so I think this is quite an important issue.

Thank you in advance.

**Comment by sujin0529** [comment]
Hi,
You could assume that grammar is given as a subset of rules.
And you should't have to consider other nonterminal except Start, Bexp.
Thanks.

**Comment by park-sunho** [comment]
Thank you so much.
